storage/local/gorilla.go

```diff
@@ -0,0 +1,1103 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package local
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/prometheus/common/model"
+)
+
+// Gorilla chunk encoding is inspired by the following paper:
+//   Gorilla: A Fast, Scalable, In-Memory Time Series Database
+//   T. Pelkonen et al., Facebook Inc.
+//   http://www.vldb.org/pvldb/vol8/p1816-teller.pdf
+// Note that there are significant differences in detail, some of which due to
+// the way Prometheus chunks work, some of which to optimize for the Prometheus
+// use-case.
+//
+// Layout of a 1024 byte gorilla chunk (big endian, wherever it matters):
```

Comment:
Most users are going to be on little endian, shouldn't we save a tad of cpu by storing things that way?